Nightly Dotty workflow of 2022-04-28 failed

See https://github.com/lampepfl/dotty/actions/runs/2236991940

Ah, our old friend the Windows directory separator.

In `test_windows_full`:

```
Error:  Test dotty.tools.languageserver.HoverTest.topLevel failed: java.lang.AssertionError: 
Error:  
Error:  // project299\Source0.scala in project project299
Error:  package hello
Error:  val x: Int = 1
Error:  val y = this.x
Error:  
Error:  while executing action: CodeHover([start=(m1,line=2,char=8), end=(m2,line=2,char=12)], Some(```scala
Error:  (hello.Source0$package : hello.Source0$package.type)
Error:  ```))
Error:  
Error:            expected:<```scala
Error:  ([hello.Source0$package : hello.Source0$package.type])
Error:  ```> but was:<```scala
Error:  ([
Error:    hello.C:\actions-runner2\_work\dotty\dotty\out\ide-tests\src\project299\Source0$package
Error:   : 
Error:    hello.
Error:      C:\actions-runner2\_work\dotty\dotty\out\ide-tests\src\project299\Source0$package
Error:    .type
Error:  ])
Error:  ```>, took 0.047 sec
```

https://github.com/lampepfl/dotty/blob/b1f00a78eba1eb22c4b2defaa88acc6e025ef59a/compiler/src/dotty/tools/io/VirtualFile.scala#L107-L109

--- a/compiler/src/dotty/tools/dotc/Run.scala
+++ b/compiler/src/dotty/tools/dotc/Run.scala
@@ -11,7 +11,7 @@ import Denotations.Denotation
 import typer.Typer
 import typer.ImportInfo.withRootImports
 import Decorators._
-import io.{AbstractFile, VirtualFile}
+import io.AbstractFile
 import Phases.unfusedPhases
 
 import util._

--- a/compiler/src/dotty/tools/dotc/interactive/InteractiveDriver.scala
+++ b/compiler/src/dotty/tools/dotc/interactive/InteractiveDriver.scala
@@ -14,7 +14,7 @@ import java.util.zip._
 import scala.collection._
 import scala.io.Codec
 
-import dotty.tools.io.{ AbstractFile, VirtualFile }
+import dotty.tools.io.AbstractFile
 
 import ast.{Trees, tpd}
 import core._, core.Decorators._

--- a/compiler/src/dotty/tools/dotc/util/SourceFile.scala
+++ b/compiler/src/dotty/tools/dotc/util/SourceFile.scala
@@ -15,6 +15,7 @@ import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 import scala.util.chaining.given
 
+import java.io.File.separator
 import java.nio.charset.StandardCharsets
 import java.nio.file.{FileSystemException, NoSuchFileException}
 import java.util.Optional
@@ -217,8 +218,11 @@ object SourceFile {
 
   implicit def fromContext(using Context): SourceFile = ctx.source
 
+  /** A source file with an underlying virtual file. The name is taken as a file system path
+   *  with the local separator converted to "/". The last element of the path will be the simple name of the file.
+   */
   def virtual(name: String, content: String, maybeIncomplete: Boolean = false) =
-    SourceFile(new VirtualFile(name, content.getBytes(StandardCharsets.UTF_8)), content.toCharArray)
+    SourceFile(new VirtualFile(name.replace(separator, "/"), content.getBytes(StandardCharsets.UTF_8)), content.toCharArray)
       .tap(_._maybeInComplete = maybeIncomplete)
 
   /** Returns the relative path of `source` within the `reference` path

--- a/compiler/src/dotty/tools/io/VirtualDirectory.scala
+++ b/compiler/src/dotty/tools/io/VirtualDirectory.scala
@@ -56,7 +56,7 @@ extends AbstractFile {
 
   override def fileNamed(name: String): AbstractFile =
     Option(lookupName(name, directory = false)) getOrElse {
-      val newFile = new VirtualFile(name, path + '/' + name)
+      val newFile = new VirtualFile(name, s"$path/$name")
       files(name) = newFile
       newFile
     }

--- a/scaladoc/src/dotty/tools/scaladoc/DocContext.scala
+++ b/scaladoc/src/dotty/tools/scaladoc/DocContext.scala
@@ -8,7 +8,6 @@ import java.nio.file.Paths
 import collection.JavaConverters._
 import dotty.tools.scaladoc.site.StaticSiteContext
 import dotty.tools.dotc.core.Contexts._
-import dotty.tools.io.VirtualFile
 import dotty.tools.dotc.util.SourceFile
 import dotty.tools.dotc.util.SourcePosition
 import dotty.tools.dotc.util.Spans

